File: Assets/zfoocs/Common/TripleString.cs
using System;
using System.Collections.Generic;
namespace zfoocs
{
    
    public class TripleString
    {
        public string left;
        public string middle;
        public string right;
    }

    public class TripleStringRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 115;
        }
    
        public void Write(ByteBuffer buffer, object packet)
        {
            if (packet == null)
            {
                buffer.WriteInt(0);
                return;
            }
            TripleString message = (TripleString) packet;
            buffer.WriteInt(-1);
            buffer.WriteString(message.left);
            buffer.WriteString(message.middle);
            buffer.WriteString(message.right);
        }
    
        public object Read(ByteBuffer buffer)
        {
            int length = buffer.ReadInt();
            if (length == 0)
            {
                return null;
            }
            int beforeReadIndex = buffer.GetReadOffset();
            TripleString packet = new TripleString();
            string result0 = buffer.ReadString();
            packet.left = result0;
            string result1 = buffer.ReadString();
            packet.middle = result1;
            string result2 = buffer.ReadString();
            packet.right = result2;
            if (length > 0)
            {
                buffer.SetReadOffset(beforeReadIndex + length);
            }
            return packet;
        }
    }
}

File: Assets/zfoolua/common/TripleString.lua

local TripleString = {}

function TripleString:new()
    local obj = {
        left = "", -- string
        middle = "", -- string
        right = "" -- string
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function TripleString:protocolId()
    return 115
end

function TripleString:protocolName()
    return TripleString
end

function TripleString:__tostring()
    local jsonTemplate = "{left:%s, middle:%s, right:%s}"
    local result = string.format(jsonTemplate, self.left, self.middle, self.right)
    return result
end

function TripleString:write(buffer, packet)
    if packet == nil then
        buffer:writeInt(0)
        return
    end
    buffer:writeInt(-1)
    buffer:writeString(packet.left)
    buffer:writeString(packet.middle)
    buffer:writeString(packet.right)
end

function TripleString:read(buffer)
    local length = buffer:readInt()
    if length == 0 then
        return nil
    end
    local beforeReadIndex = buffer:getReadOffset()
    local packet = TripleString:new()
    local result0 = buffer:readString()
    packet.left = result0
    local result1 = buffer:readString()
    packet.middle = result1
    local result2 = buffer:readString()
    packet.right = result2
    if length > 0 then
        buffer:setReadOffset(beforeReadIndex + length)
    end
    return packet
end

return TripleString
